src/fn/move-contacts.js

```diff
@@ -0,0 +1,273 @@
+const minimist = require('minimist');
+const path = require('path');
+const readline = require('readline-sync');
+
+const fs = require('../lib/sync-fs');
+const log = require('../lib/log');
+const pouch = require('../lib/db');
+const lineageManipulation = require('../lib/lineage-manipulation');
+const lineageConstraints = require('../lib/lineage-constraints');
+
+const { warn, trace, info, error } = log;
+
+const HIERARCHY_ROOT = 'root';
+
+module.exports = (projectDir, couchUrl, extraArgs) => {
+  const args = parseExtraArgs(projectDir, extraArgs);
+  const db = connectToDatabase(couchUrl);
+  prepareDocumentDirectory(args);
+  return updateLineagesAndStage(args, db);
+};
+
+const prettyPrintDocument = doc => `'${doc.name}' (${doc._id})`;
+const updateLineagesAndStage = async (options, db) => {
+  trace(`Fetching contact details for parent: ${options.parentId}`);
+  const parentDoc = await fetch.contact(db, options.parentId);
+
+  const constraints = await lineageConstraints(db, parentDoc);
+  const contactDocs = await fetch.contactList(db, options.contactIds);
+  await preprocessContacts(contactDocs, constraints);
+
+  let affectedContactCount = 0, affectedReportCount = 0;
+  const replacementLineage = lineageManipulation.createLineageFromDoc(parentDoc);
+  for (let contactId of options.contactIds) {
+    const contactDoc = contactDocs[contactId];
+    const descendantsAndSelf = await fetch.descendantsOf(db, contactId);
+    
+    // Check that primary contact is not removed from areas where they are required
+    const invalidPrimaryContactDoc = await constraints.getPrimaryContactViolations(contactDoc, descendantsAndSelf);
+    if (invalidPrimaryContactDoc) {
+      throw Error(`Cannot remove contact ${prettyPrintDocument(invalidPrimaryContactDoc)} from the hierarchy for which they are a primary contact.`);
+    }
+
+    trace(`Considering lineage updates to ${descendantsAndSelf.length} descendant(s) of contact ${prettyPrintDocument(contactDoc)}.`);
+    const updatedDescendants = replaceLineageInContacts(descendantsAndSelf, replacementLineage, contactId);
+    
+    const ancestors = await fetch.ancestorsOf(db, contactDoc);
+    trace(`Considering primary contact updates to ${ancestors.length} ancestor(s) of contact ${prettyPrintDocument(contactDoc)}.`);
+    const updatedAncestors = replaceLineageInAncestors(descendantsAndSelf, ancestors);
+
+    const reportsCreatedByDescendants = await fetch.reportsCreatedBy(db, descendantsAndSelf.map(descendant => descendant._id));
+    trace(`${reportsCreatedByDescendants.length} report(s) created by these affected contact(s) will update`);
+    const updatedReports = replaceLineageInReports(reportsCreatedByDescendants, replacementLineage, contactId);
+    
+    [...updatedDescendants, ...updatedReports, ...updatedAncestors].forEach(updatedDoc => {
+      lineageManipulation.minifyLineagesInDoc(updatedDoc);
+      writeDocumentToDisk(options, updatedDoc);
```

Comment:
Consider collecting all of these and writing them to disk once all the docs are processed. 
This way you won't end up with a "partial" update (if any of the contacts throws) that one can still upload.

Reply:
I think that if this thing throws, nobody should be uploading the result. That sounds dangerous.